File: javascript/dominoes/dominoes.js
//
// This is only a SKELETON file for the 'Dominoes' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const chain = (tiles) => {
  if (tiles.length === 0)
    return []
  // Start the chain with the first tile.
  let chained = [tiles[0]]
  let result = helper(chained, tiles.slice(1))
  if (result != null && result.length !== tiles.length)
  {
    // If that didn't work, try again with the tile reversed.
    chained = [tileReverse(tiles[0])]
    result = helper(chained, tiles.slice(1))
  }
  if (result == null || result.length !== tiles.length)
    return null

  if (result[0][0] !== result[result.length - 1][1])
    return null
  return result
  
};

function helper(chained, remaining)
{
  if (remaining.length === 0)
    return chained

  let first = chained[0][0]
  let last = chained[chained.length - 1][1]

  // last one?
  if (remaining.length === 1)
  {
    let lastTile = remaining[0]
    if (lastTile[1] === first && lastTile[0] === last)
    {
      // match chain ends when normal
      return [...chained, lastTile]
    }
    else if (lastTile[0] === first && lastTile[1] === last)
    {
      // match chain ends when reversed
      return [...chained, tileReverse(lastTile)]
    }
    else
      return null
  }

  for(let i = 0; i < remaining.length; i++)
  {
    let current = remaining[i]
    let currentReverse = tileReverse(current)

    if (current[1] === first)
    {
      // match first tile 
      let result = helper([current, ...chained], [...remaining.filter((value, index) => index !== i)])
      if (result !== null && result.length === chained.length + remaining.length)
        return result      
    }
    else if (currentReverse[1] === first)
    {
      // match first tile reversed
      let result = helper([currentReverse, ...chained], [...remaining.filter((value, index) => index !== i)])
      if (result !== null && result.length === chained.length + remaining.length)
        return result      
    }
    else if (current[0] === last)
    {
      // match last tile 
      let result = helper([...chained, current], [...remaining.filter((value, index) => index !== i)])
      if (result !== null && result.length === chained.length + remaining.length)
        return result      
    }
    else if (currentReverse[0] === last)
    {
      // match last tile 
      let result = helper([...chained, currentReverse], [...remaining.filter((value, index) => index !== i)])
      if (result !== null && result.length === chained.length + remaining.length)
        return result      
    }
  }

  return null // No matches for current chain
}

function tileReverse(tile)
{
  return [tile[1], tile[0]]
}

File: python/dominoes/dominoes.py
"""
Chain a set of tiles into a loop of dominoes with matching number linearly and end to end
"""

def can_chain(dominoes):
    """Check if a set of dominoes can be chained in a circular fashion
    :param dominoes: list of length two tuples where two numbers represent the numbers on the top and bottom of a dominoe
    :returns: None if no pattern found. Otherwise the dominoes arranged where they match linearly and at the ends.
    """
    if len(dominoes) == 0:
        return []
    # Start the chain with the first tile.
    chained = [dominoes[0]]
    result = helper(chained, dominoes[1:])
    if result is not None and len(result) != len(dominoes):  
        # If that didn't work, try again with the tile reversed.
        chained = [tile_reverse(dominoes[0])]
        result = helper(chained, dominoes[1:])

    if result is None or len(result) != len(dominoes):
        return None

    if result[0][0] != result[len(result) - 1][1]:
        return None
    return result

def helper(chained, remaining):
    """
    Recursive helper function used to find a matching chain of dominoes
    :param chained: The list of dominoes chained so far
    :param remaining: The list of dominoes not yet in the chain
    :returns: None if no chain found, otherwise a list of dominoes that can be chained linearly and end to end.
    """
    def result_ok(result):
        return result is not None and len(result) == len(chained) + len(remaining)
    
    def match_first(tile, target, chained, remaining):
        reverse_tile = tile_reverse(tile)
        if current[1] == target:
            result = helper([tile, *chained], remaining)
            if result_ok(result):
                return result
        if reverse_tile[1] == target:
            result = helper([reverse_tile, *chained], remaining)
            if result_ok(result):
                return result
        return None
    
    def match_last(tile, target, chained, remaining):
        reverse_tile = tile_reverse(tile)
        if current[0] == target:
            result = helper([*chained, tile], remaining)
            if result_ok(result):
                return result
        if reverse_tile[0] == target:
            result = helper([*chained, reverse_tile], remaining)
            if result_ok(result):
                return result
        return None
    
    if len(remaining) == 0:
        return chained
    
    first = chained[0][0]
    last = chained[len(chained) - 1][1]

    # last one?
    if len(remaining) == 1:
        last_tile = remaining[0]
        if last_tile[1] == first and last_tile[0] == last:
            # match chain ends when normal
            return [*chained, last_tile]
        if last_tile[0] == first and last_tile[1] == last:
            # match chain ends when reversed
            return [*chained, tile_reverse(last_tile)]
        return None
    
    for index, current in enumerate(remaining):
        remaining_without_current = [tile for tile_index, tile in enumerate(remaining) if tile_index != index]
        
        result = match_first(current, first, chained, remaining_without_current)
        if result_ok(result):
            return result
        result = match_last(current, last, chained, remaining_without_current)
        if result_ok(result):
            return result
        
    return None # No matches for current chain

def tile_reverse(tile):
    """
    Reverse a single dominoe (a, b) => (b, a)
    :param tile: a tuple of two numbers representing a single dominoe
    :returns: a tuple of two numbers representing a dominoe in the reverse order of the original tile (as if flipped around 180 degrees)
    """
    return [tile[1], tile[0]]
